File: Estudo Dirigido/Experimentos/Tree.cpp
#include <iostream>
#include <vector>

using namespace std;

//para essa formulação, consideramos uma árvore binária perfeita de profundidade d
class Tree{

    private: 
        int depth;
        vector<int> nodes;
        vector<int> leaves;

    public:
        //construtor
        Tree(int d) : depth(d) {
            //nós internos (de branch): [1, ..., 2^depth - 1]
            for (int i = 1; i <= (1 << depth) - 1; i++) {
                nodes.push_back(i);
            }
            //nós folha: [2^depth, ..., 2^(depth + 1) - 1]
            for (int i = (1 << depth); i <= (1 << (depth + 1)) - 1; i++) {
                leaves.push_back(i);
            }
        }

        int getDepth() {
            return depth;
        }

        int leftChild(int node) {
            return 2 * node;
        }

        int rightChild(int node) {
            return 2 * node + 1;
        }

        const vector<int>& getNodes() const {
            return nodes;
        }

        const vector<int>& getLeaves() const {
            return leaves;
        }
};

File: Estudo Dirigido/Experimentos/Tree.py
#!/usr/bin/env python3
class Tree:

    def __init__(self, depth):
        self.depth = depth
        #nós de branch: [1, ..., 2^depth - 1]
        self.nodes = [i for i in range(1, (1 << depth))]
        # nós folha: [2^depth, ..., 2^(depth + 1) - 1]
        self.leaves = [i for i in range((1 << depth), (1 << (depth + 1)))]

    def get_depth(self):
        return self.depth

    def left(self, node):
        return 2 * node

    def right(self, node):
        return 2 * node + 1

    def get_nodes(self):
        return self.nodes

    def get_leaves(self):
        return self.leaves
    
    def get_parent(self, node):
        return node // 2
